### DISCUSSION TITLE
Oracle Support

### DISCUSSION THREAD

**Discussion by angkanas** [fastapi/sqlmodel#641] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from sqlmodel import Field, Session, SQLModel, create_engine


class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: Optional[int] = None

engine = create_engine("oracle+cx_oracle://<user>:<password>@<dbserver>")


SQLModel.metadata.create_all(engine)
```

### Description

sqlalchemy.exc.DatabaseError: (cx_Oracle:Database:Error) ORA-00906: missing left parenthesis
[SQL:
CREATE TABLE hero(
  id INTERGER NOT NULL,
  name VARCHAR2 NOT NULL,
  secret_name VARCHAR2 NOT NULL,
  age INTERGER,
  PRIMARY KEY(id)
)
]

### Operating System

Windows

### SQLModel Version

0.0.8

### Python Version

3.11.4

**Comment by spartyhacker** [comment]
I had exactly the same issue. It is due to the fact that VARCHAR2 needs a length spec. However, I was not able to figure out how to use sqlmodel to do that. The workaround I found is: create the table in Oracle before using sqlmodel. That way the "create table" sql is not implemented hence avoided the error

**Comment by SonOfLilit** [comment]
Are you sure this is an SQLModel issue? It really looks like an SQLAlchemy issue. Can you show working SQLAlchemy code similar to this?

**Reply by sandstromviktor** [reply]
```python
# WORKS (SQLAlchemy)
class MyTable(Base):
    __tablename__ = 'my_table'

    id = Column(Integer, primary_key=True)
    name = Column(String(10))  # Specify length here

# DOES NOT WORK (generates VARCHAR2 and not VARCHAR2(10 CHAR))
class MyTable(SQLModel):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str = Field(sa_column=String(10))
```

**Reply by sandstromviktor** [reply]
I found a solution. Described in #1133

**Comment by AdityaPunetha** [comment]
I'm facing the same issue, i have used the solution described in https://github.com/fastapi/sqlmodel/issues/1133
can we at least create a open github issue so someone or me could fix it

edit- I'm also facing issues where auto increment is not working with oracle, when i try to insert a new row without passing the id explicitly it throws a null value error

**Reply by YuriiMotov** [reply]
Could you please clarify what your suggestion is?

In the code example above, it's wrong to pass `String(10)` to `sa_column`. This parameter accepts `Column` object, so, we should pass `sa_column=Column(String(10))`. Could you please try it?

The solution with `max_length` also looks correct to me.

**Reply by AdityaPunetha** [reply]
If you look at the documentation, we define a column type by simply declaring the type without any additional attributes:
```
class Hero(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: int | None = None
```
However, this approach doesn’t work with OracleDB. We should maintain consistency in such cases otherwise, it will create issues during database migration. Changing the database shouldn’t require restructuring the code; that’s the main purpose of using an ORM to provide a consistent interface and eliminate the need to remember different SQL syntaxes for different databases.

If the examples provided in the official documentation don’t actually work across databases, it raises a serious question what’s the point of using this library over SQLAlchemy in the first place? At the very least, the examples in the documentation should work as expected

Additionally, the error message I received was quite unclear. It only stated “missing left parenthesis,” which didn’t help in diagnosing or understanding the actual issue. Ideally, such errors should provide more context or guidance. In this case, I shouldn’t have needed to look up solutions for something that appears exactly as it does in the documentation.

**Reply by YuriiMotov** [reply]
If user doesn't specify the length of the string, what should we use to create the column?
Following the idea of making it consistent across different databases, we will have to set the length for all string columns. It will be a breaking change and unexpected for those who uses databases where the length is not required.

> Changing the database shouldn’t require restructuring the code; that’s the main purpose of using an ORM to provide a consistent interface and eliminate the need to remember different SQL syntaxes for different databases.

I agree that would be ideal, but.. 
I think that ORM provides you a way to create DB-agnostic code, but you should still adapt (in advance) this code for those databases you want this code to work with.
In this case, if you want this code be compatible with various databases, just manually specify the length of the string.
And, it's the same with SQLAlchemy - simple code will work with different databases, but not all features.

> Additionally, the error message I received was quite unclear. It only stated “missing left parenthesis,” which didn’t help in diagnosing or understanding the actual issue. 

I think this error message comes directly from the Oracle driver. I'm afraid it's possible to handle all error messages and translate them to more user-readable..
